SCP-1046: Prism prototype

j-mueller opened the pull request:
`Language.PlutusTx.Coordination.Contracts.Prism` and its submodules define the Plutus part of a credential management dapp such as Atala PRISM / Mirror. There are two actors:

* Mirror. Issues and revokes the on-chain credential tokens.
* User. Uses credentials in transactions

We have the following modules:

1. `Credential`: Defines the 'Credential' type and a monetary policy script (MPS) for creating and destroying credential tokens
2. `StateMachine`: Defines the state machine script that allows Users to present their credentials in transactions and Mirror to revoke/destroy credentials.
3. `STO`: A simple Security Token Offering (STO) that requires a credential to be presented by anyone who wants to participate. This is provided for testing and demo purposes
4. `CredentialManager`: A Plutus application managing all the credentials that are available to the User.
5. `Mirror`: A Plutus application that locks credentials in state machines and revokes them when necessary.
6. `Withdraw`: A Plutus application that asks the 'CredentialManager' app for  credentials and then participates in the STO.

There is a test in `Spec.Prism` that runs through the whole process.

------

I made some modifications to `Language.Plutus.Contract.StateMachine`, making it easier to get the data needed to transition a state machine instance. This is useful if you want to run other scripts in the same transaction.

michaelpj commented on plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Prism/StateMachine.hs:
```diff
@@ -0,0 +1,120 @@
+
+data IDState =
+    Active -- ^ The credential is active and can be used in transactions
+    | Revoked -- ^ The credential has been revoked and can't be used anymore.
```
If it's a state machine per token (which it seems to be), why not just terminate the machine when it's revoked?

michaelpj replied:
Ah, that is what happens.

michaelpj commented on plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Prism/StateMachine.hs:
```diff
@@ -0,0 +1,120 @@
+        , ucToken      :: Value
+        -- ^ The 'Value' containing a token of the credential
+        -- (this needs to be included here because 'Credential.token'
+        -- is not available in on-chain code)
```
This feels suspicious, need to think about it.

j-mueller replied:
It's a common issue, see for example the [`FutureAccounts`](https://github.com/input-output-hk/plutus/blob/master/plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs#L116) datatype from the futures contract

michaelpj commented on plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Prism/StateMachine.hs:
```diff
@@ -0,0 +1,120 @@
+    let val = $$(PlutusTx.compile [|| validator ||]) `PlutusTx.applyCode` PlutusTx.liftCode credentialData
+        validator d = StateMachine.mkValidator (credentialStateMachine d)
+        wrap = Scripts.wrapValidator @IDState @IDAction
+    in Scripts.validator @(StateMachine IDState IDAction) val $$(PlutusTx.compile [|| wrap ||])
```
Ugh this boilerplate still hurts me every time I see it :(

michaelpj commented on plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Prism/Withdraw.hs:
```diff
@@ -0,0 +1,108 @@
+{-# LANGUAGE TypeApplications   #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# LANGUAGE TypeOperators      #-}
+-- | The withdrawal application that withdraws some funds
```
from where?

michaelpj replied:
Might be nice to do the exchange withdrawal example too? Which is even simpler: just a way to lock an output in a way that you can only unlock it when presenting a particular token.

j-mueller replied:
Yes, I just remembered that we already have a `TokenAccount` contract that does exactly this.

michaelpj reviewed (approved):
Generally looks good! Could do with a bit more haddock.

j-mueller commented on plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Prism/STO.hs:
```diff
@@ -0,0 +1,75 @@
+validateSTO STOData{stoIssuer,stoCredentialToken,stoTokenName} PolicyCtx{policyCtxTxInfo=txInfo,policyCtxPolicy=ownHash} =
+    let tokenOK = stoCredentialToken `Value.leq` Validation.valueSpent txInfo
+        Lovelace paidToIssuer = fromValue (Validation.valuePaidTo txInfo stoIssuer)
+        forgeOK =
```
True. They would just be worthless tokens though.

j-mueller commented on plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:
```diff
@@ -253,7 +255,15 @@ runInitialise StateMachineClient{scInstance} initialState initialValue = mapErro
-type StateMachineTypedTx state input = (TxConstraints Void Void, State state, State state, InputConstraint input, ScriptLookups (StateMachine state input))
+
+-- | Constraints & lookups needed to transition a state machine instance
+data StateMachineTypedTxContext state input =
```
Renamed it to `StateMachineTransition`...

j-mueller commented on plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Prism.hs:
```diff
@@ -0,0 +1,97 @@
+
+import           Language.Plutus.Contract
+
+-- $prism
```
I gave up and just put everything in a multiline comment at the top of the module. Also extended the comment a bit.

j-mueller commented on plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Prism/Mirror.hs:
```diff
@@ -0,0 +1,92 @@
+
+type MirrorSchema =
+    BlockchainActions
+        .\/ Endpoint "issue" (TokenName, PubKeyHash) -- lock a single token in a state machine tied to the client address
```
I named it `CredentialOwnerReference`

j-mueller commented on plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Prism/Mirror.hs:
```diff
@@ -0,0 +1,92 @@
+    => Contract s MirrorError ()
+mirror = do
+    authority <- mapError SetupError $ CredentialAuthority . pubKeyHash <$> ownPubKey
+    let go = (createTokens authority `select` revokeToken authority) >> go
```
`forever` didn't use to work in the "old" implementation of `Contract`, before I switched over to `freer-simple` effects (because it uses `Applicative`, not `Monad`, resulting in a loop). It should be fine now though, I'll check.

j-mueller replied:
Yep it works

j-mueller commented:
@michaelpj While addressing the review comments I added a thing to `Tx.Constraints.OffChain` for combining the constraints of several typed scripts: https://github.com/input-output-hk/plutus/pull/2299/commits/92ed8f546bfe02c83cf02ad63d076f25ac71a187#diff-17af04aec13259aec586db8b3fa271c0

If you're happy with that then I think we can merge this PR.
